watch-parties/src/main/java/com/google/sps/servlets/ChatListerServlet.java

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import java.util.Date;
+import com.google.sps.data.Chat;
+import java.util.LinkedList; 
+import java.util.Stack; 
+
+
+@WebServlet("/chatstorage")
+public class ChatListerServlet extends HttpServlet {
+
+  public ArrayList<Chat> comments = new ArrayList<>();
+  public Stack<Chat> stackComments = new Stack<>();
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Bring comments back from datastore
+    Query query = new Query("Chat").addSort("timestamp", SortDirection.ASCENDING); 
+    PreparedQuery results = datastore.prepare(query);
+
+    // count how many comments there are
+    int count = 0;
```

Comment:
I think there are a couple things we can do to simplify this.

(1) sort by descending order so that your query give you the most recent chats. We want the highest time stamp at the top of the list since we only care about the newest chats.

(2) use FetchOptions to limit the number of chats you fetch. https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/FetchOptions.html
You should be able to do results.asIterable(FetchOptions.Builder.withLimit(50)) and your iterator will only have the 50.

(3) You want to print the messages in ascending order, so that the oldest chat messages are at the top and the newest are at the bottom. You can either re-sort your items into ascending order or use results.asList(FetchOptions.Builder.withLimit(50)).toArray() to get an array of your Entities and then traverse that array from back to front.